packages/malloy/src/dialect/dialect.ts

```diff
@@ -276,13 +276,214 @@ export abstract class Dialect {
+   *
+   * @param expr The SQL expression to offset
+   * @param op The operation: '+' for addition, '-' for subtraction
+   * @param magnitude The SQL expression for the interval magnitude (e.g., '6', '(delta_val)')
+   * @param unit The interval unit (year, month, day, hour, etc.)
+   * @param typeDef The Malloy type of the expression (date, timestamp, etc.)
+   * @param inCivilTime If true, the expression is already in civil (local) time and should not
+   *                    be converted. If false, may need timezone conversion for timestamps.
+   * @param timezone Optional timezone for the operation. Only provided when timezone-aware
+   *                 offset is needed but inCivilTime is false.
+   * @returns SQL expression representing the offset time
+   */
+  abstract sqlOffsetTime(
+    expr: string,
+    op: '+' | '-',
+    magnitude: string,
+    unit: TimestampUnit,
+    typeDef: AtomicTypeDef,
+    inCivilTime: boolean,
+    timezone?: string
+  ): string;
+
+  /**
+   * Determines whether a truncation and/or offset operation needs to be performed in
+   * civil (local) time, and if so, which timezone to use.
+   *
+   * Calendar-based operations (day, week, month, quarter, year) typically need civil time
+   * computation because these units can cross DST boundaries, changing the UTC offset.
+   *
+   * Default implementation:
+   * - Returns true for timestamps with truncation or calendar-unit offsets
+   * - Uses the query timezone from QueryInfo, or 'UTC' if none specified
+   * - Returns false for dates or sub-day offsets (hour, minute, second)
+   *
+   * Dialects can override this if they have different rules about which operations
+   * require civil time computation.
+   *
+   * @param typeDef The Malloy type of the base expression
+   * @param truncateTo The truncation unit, if any
+   * @param offsetUnit The offset unit, if any
+   * @param qi Query information including timezone settings
+   * @returns Object with `needed` boolean and optional `tz` string
+   */
+  needsCivilTimeComputation(
+    typeDef: AtomicTypeDef,
+    truncateTo: TimestampUnit | undefined,
+    offsetUnit: TimestampUnit | undefined,
+    qi: QueryInfo
+  ): {needed: boolean; tz: string | undefined} {
+    // Calendar units that can cross DST boundaries
+    const calendarUnits = ['day', 'week', 'month', 'quarter', 'year'];
+
+    const isCalendarTruncate =
+      truncateTo !== undefined && calendarUnits.includes(truncateTo);
+
+    const isCalendarOffset =
+      offsetUnit !== undefined && calendarUnits.includes(offsetUnit);
+
+    const tz = qtz(qi);
+
+    // Timestamps with calendar truncation/offset need civil time computation
+    // BUT only if there's actually a timezone to convert to/from
+    const needed =
+      TD.isTimestamp(typeDef) &&
+      (isCalendarTruncate || isCalendarOffset) &&
+      tz !== undefined;
+
+    return {needed, tz};
+  }
+
+  /**
+   * Unified function for truncation and/or offset operations.
```

Comment:
mention that the intention is that this NOT be over-ridden and the dialect quirks should be in the helper functions.